### PR TITLE
DS-4107 Represent DSO metadata as a map in REST

### PR DESCRIPTION
**Pull request by cwilper**
REST Contract: https://github.com/DSpace/Rest7Contract/pull/39

This modifies the DSpace 7 REST API implementation to model embedded DSO metadata as maps, as described here: https://jira.duraspace.org/browse/DS-4107

Most of these changes were lifted from the POC PR: https://github.com/DSpace/DSpace/pull/2175  ...where representing metadata as a map was identified as useful keep.

Overall lines of code were reduced because the change provided an opportunity to share code in some places.

ITs have also been updated to expect the new form.

### CODE REVIEW

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MetadataValueRest.java [hunk 1]
```diff
@@ -26,7 +25,7 @@
 
     int confidence;
 
-    @JsonProperty(access = Access.READ_ONLY)
+    @JsonIgnore
```
Comment (terrywbrady) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MetadataValueRest.java]: I pause at this point in my review to wonder why this is suppressed from the JSON.  I see that there was some discussion on this in the Rest Contract.  Perhaps an additional comment would be useful here.

**Review by terrywbrady — COMMENTED** [review]
@cwilper , this is a nice cleanup of the code.  The changes look good.

I completed my review.  I see that a re-base of the code is needed.  Ping me once that is resolved and I will run a quick test.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataConverter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.dspace.app.rest.model.MetadataValueRest;
+import org.dspace.content.DSpaceObject;
+import org.dspace.content.MetadataValue;
+import org.dspace.content.service.DSpaceObjectService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.stereotype.Component;
+
+@Component
+public class MetadataConverter implements Converter<List<MetadataValue>, MetadataRest> {
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataConverter.java]: Please add javadoc here

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataConverter.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        // Populate MetadataRest's map of key -> List while respecting SortedSet's order
+        Map<String, List<MetadataValueRest>> mapOfLists = metadataRest.getMap();
+        for (Map.Entry<String, SortedSet<MetadataValueRest>> entry : mapOfSortedSets.entrySet()) {
+            mapOfLists.put(entry.getKey(), entry.getValue().stream().collect(Collectors.toList()));
+        }
+
+        return metadataRest;
+    }
+
+    public void setMetadata(DSpaceObjectService dsoService, DSpaceObject dso, MetadataRest metadataRest) {
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataConverter.java]: should this be a private method. Javadoc here will be appreciated as well

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataValueConverter.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.dspace.app.rest.converter;
+
+import org.dspace.app.rest.model.MetadataValueRest;
+import org.dspace.content.MetadataValue;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.stereotype.Component;
+
+@Component
+public class MetadataValueConverter implements Converter<MetadataValue, MetadataValueRest> {
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataValueConverter.java]: please add basic javadoc

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DSpaceObjectRest.java [hunk 5]
```diff
@@ -53,11 +51,11 @@ public void setHandle(String handle) {
         this.handle = handle;
     }
 
-    public List<MetadataEntryRest> getMetadata() {
+    public MetadataRest getMetadata() {
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DSpaceObjectRest.java]: please add javadoc for all public methods

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MetadataRest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+
+public class MetadataRest {
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MetadataRest.java]: please add javadoc here

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java [hunk 7]
```diff
@@ -407,4 +418,28 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
         throw new RepositoryMethodNotImplementedException("No implementation found; Method not allowed!", "");
     }
 
+    protected void updateMetadata(DSpaceObjectService dsoService,
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: please add javadoc here

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/matcher/MetadataMatcher.java [hunk 8]
```diff
@@ -8,19 +8,20 @@
...
 
 import static com.jayway.jsonpath.matchers.JsonPathMatchers.hasJsonPath;
-import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.is;
 
 import org.hamcrest.Matcher;
 
-public class CommunityMetadataMatcher {
+public class MetadataMatcher {
```
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/matcher/MetadataMatcher.java]: please add javadoc here

**Review by abollini — CHANGES_REQUESTED** [review]
Hi @cwilper thanks for this PR: the implementation looks fine to me. Please add the necessary javadoc for public method and solve the conflict so that we can formally approve the PR. This PR will have a big impact on the angular side so I recommend to get a correspondent PR on the angular side ready before to merge

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java (hunk 7 above)
Comment (cwilper) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: N/A - method dropped in favor of MetadataConverter#setMetadata in latest rebase

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataConverter.java (hunk 3 above)
Comment (cwilper) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/MetadataConverter.java]: Added javadocs to explain purpose of this method and how it is basically the inverse of convert method in the same class -- it converts in the opposite direction: from rest to domain object. It is used by the repositories and needs to be public.

**Comment by cwilper** [comment]
Thanks for taking a look @terrywbrady and @abollini . I have rebased on master and believe I have implemented all suggested improvements (mainly javadocs).

Note that the angular side of this was done a few weeks ago. I will shortly be rebasing that as well, after which I will link here. I do **recommend this only be merged when the related angular changes are also ready to be merged**.

**Review by abollini — APPROVED** [review]
thanks for the javadoc. It is fine for me now

**Comment by cwilper** [comment]
Angular implementation has been rebased and is ready for review: https://github.com/DSpace/dspace-angular/pull/347

**Review by terrywbrady — APPROVED** [review]
The changes look good.  I did a quick test in HAL and I see that the map is present for collections and items.

**Comment by cwilper** [comment]
Rebased to latest master

**Comment by tdonohue** [comment]
Merging as this is at +2, and looks good to me too!

**Pull request merged**